i6_native_ops/fbw/__init__.py

```diff
@@ -1,2 +1,65 @@
 import torch  # needed to find pytorch specific libs
-from .fbw_core import fbw
+from pkg_resources import get_distribution
+
+try:
+    # Package is installed, so ops are already compiled
+    __version__ = get_distribution('i6_native_ops').version
+    from . import fbw_core as core
+except Exception as e:
+    # otherwise try to build locally
+    from torch.utils.cpp_extension import load
+    base_path = os.path.dirname(__file__)
+    core = load(
+        name="fbw_core",
+        sources=[
+            os.path.join(base_path, "fbw_torch.cpp"),
+            os.path.join(base_path, "fbw_op.cu"),
+        ],
+        extra_include_paths=[
+            base_path,
+            os.path.join(base_path, "..", "common")
+        ],
+    )
+
+
+class FastBaumWelchLoss(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, am_scores, fsa, seq_lens):
+        num_states, edge_tensor, weight_tensor, start_end_states = fsa
+
+        grad, loss = fbw(
+            am_scores, edge_tensor, weight_tensor,
+            start_end_states, seq_lens, int(num_states),
+            nativeops.DebugOptions()
+        )
+        ctx.save_for_backward(grad)
+        return loss
+    
+    @staticmethod
+    def backward(ctx, grad_loss):
+        # negative log prob -> prob
+        grad = ctx.saved_tensors[0].neg().exp()
+        return grad, None, None
+
+
+def fbw_loss(
+    neg_log_probs: torch.FloatTensor,
+    fsa: Tuple[int, torch.IntTensor, torch.FloatTensor, torch.IntTensor],
+    seq_lens: torch.IntTensor
+) -> Tuple[torch.IntTensor, torch.FloatTensor]:
+    """ 
+    Computes negative log likelihood of an emission model given an HMM finite state automaton.
+    The corresponding gradient with respect to the emission model is automatically backpropagated.
+    :param log_probs: negative log probabilities of emission model as a (T, B, F)
```

Comment:
```suggestion
    :param neg_log_probs: negative log probabilities of emission model as a (T, B, F)
```
to be consistent with actual parameter name.

However, would it be more consistent to other ops/losses in PT if instead the (positive) log_probs were being used? We could then in this function multiply with -1 if negative scores were required.

Reply:
Yes I agree that accepting positive log probabilities would be less confusing here.